File: cpp/hardware/power/PowerSupply.cpp
#pragma once
#include <bits/stdc++.h>

#include "../../component/Component.cpp"
#include "../../component/Hardware.cpp"
#include "../../interfaces/Upgradable.cpp"

class PowerSupply : public Hardware, public Component
{
private:
    int wattage;
    std::string certification;
    bool modular;

public:
    // Constructors
    PowerSupply() : Hardware(), Component(), wattage(0), certification(""), modular(false) {}
    PowerSupply(int wattage, const std::string &certification, bool modular, const std::string &id, const std::string &manufacturer, const std::string &model, double price, const std::string &serialNumber, const Date &installationDate, bool physicallyInstalled)
        : Hardware(serialNumber, installationDate, physicallyInstalled), Component(id, manufacturer, model, price), wattage(wattage), certification(certification), modular(modular) {}

    // Destructor
    ~PowerSupply() {}

    // Setters
    void setWattage(int wattage)
    {
        this->wattage = wattage;
    }

    void setCertification(const std::string &certification)
    {
        this->certification = certification;
    }

    void setModular(bool modular)
    {
        this->modular = modular;
    }

    // Getters
    int getWattage() const
    {
        return wattage;
    }

    std::string getCertification() const
    {
        return certification;
    }

    bool isModular() const
    {
        return modular;
    }
};

File: python/hardware/power/power_supply.py
from datetime import date
from components.component import Component
from components.hardware import Hardware

class PowerSupply(Hardware, Component):
    def __init__(self, wattage=0, certification="", modular=False, id="", manufacturer="", model="", price=0.0, serial_number="", installation_date=None, physically_installed=False):
        Hardware.__init__(self, serial_number, installation_date, physically_installed)
        Component.__init__(self, id, manufacturer, model, price)
        self._wattage = wattage
        self._certification = certification
        self._modular = modular

    # Setters
    def set_wattage(self, wattage):
        self._wattage = wattage

    def set_certification(self, certification):
        self._certification = certification

    def set_modular(self, modular):
        self._modular = modular

    # Getters
    def get_wattage(self):
        return self._wattage

    def get_certification(self):
        return self._certification

    def is_modular(self):
        return self._modular